kv/txn.go

```diff
@@ -17,23 +17,112 @@ package kv
 import (
 	"context"
 	"errors"
+	"fmt"
 	"math"
 	"math/rand"
+	"sync"
 	"time"
 
 	"github.com/pingcap/failpoint"
 	"github.com/pingcap/tidb/parser/terror"
 	"github.com/pingcap/tidb/util/logutil"
+	"github.com/tikv/client-go/v2/oracle"
 	"go.uber.org/zap"
 )
 
+const (
+	// TimeToPrintLongTimeInternalTxn is the duration if the internal transaction lasts more than it,
+	// TiDB prints a log message
+	TimeToPrintLongTimeInternalTxn = time.Minute * 5
+)
+
+var globalInnerTxnTsBox = innerTxnStartTsBox{
+	innerTSLock:        sync.Mutex{},
+	innerTxnStartTsMap: make(map[uint64]struct{}, 256),
+}
+
+type innerTxnStartTsBox struct {
+	innerTSLock        sync.Mutex
+	innerTxnStartTsMap map[uint64]struct{}
+}
+
+// wrapStoreInterTxnTS is the entry function used to store the startTS of an internal transaction to globalInnerTxnTsBox
+// @param startTS	the startTS of the internal transaction produced by `RunInNewTxn`
+func wrapStoreInterTxnTS(startTS uint64) {
+	globalInnerTxnTsBox.storeInnerTxnTS(startTS)
+}
+
+func (ib *innerTxnStartTsBox) storeInnerTxnTS(startTS uint64) {
+	ib.innerTSLock.Lock()
+	ib.innerTxnStartTsMap[startTS] = struct{}{}
+	ib.innerTSLock.Unlock()
+
+}
+
+// wrapDeleteInterTxnTS delete the startTS from globalInnerTxnTsBox when the transaciotn is commted
+// @param startTS	the startTS of the internal transaction produced by `RunInNewTxn`
+func wrapDeleteInterTxnTS(startTS uint64) {
+	globalInnerTxnTsBox.deleteInnerTxnTS(startTS)
+}
```

Comment:
Do we still need the function? Its content is too simple...

Reply:
I‘ll remove it